sw/device/lib/dif/dif_aes_unittest.cc

```diff
@@ -328,122 +397,365 @@ TEST_F(Data, DataOut) {
+                     AES_CTRL_SHADOWED_MODE_VALUE_AES_ECB,
+                     AES_CTRL_SHADOWED_OPERATION_VALUE_AES_ENC,
+                     AES_CTRL_SHADOWED_PRNG_RESEED_RATE_VALUE_PER_1,
+                     .manual_op = false, .force_zero_masks = false,
+                     .key_sideloaded = false});
+
+  transaction.reseed_on_key_change = true;
+  transaction.reseed_on_key_change_lock = true;
+
+  EXPECT_READ32(AES_CTRL_AUX_REGWEN_REG_OFFSET,
+                {{AES_CTRL_AUX_REGWEN_CTRL_AUX_REGWEN_BIT, true}});
+  EXPECT_READ32(AES_CTRL_AUX_SHADOWED_REG_OFFSET,
+                {{AES_CTRL_AUX_SHADOWED_KEY_TOUCH_FORCES_RESEED_BIT, false}});
+
+  EXPECT_EQ(dif_aes_start(&aes_, &transaction, &kKey_, NULL), kDifError);
+}
 
+// Dif functions.
+class DifFunctionsTest : public AesTestInitialized {
+ protected:
+  DifFunctionsTest() {}
+};
+
+TEST_F(DifFunctionsTest, reset) {
+  EXPECT_READ32(AES_STATUS_REG_OFFSET, 1);
+  EXPECT_WRITE32_SHADOWED(AES_CTRL_SHADOWED_REG_OFFSET,
+                          {{AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, true}});
+
+  EXPECT_WRITE32(AES_TRIGGER_REG_OFFSET,
+                 {
+                     {AES_TRIGGER_KEY_IV_DATA_IN_CLEAR_BIT, true},
+                     {AES_TRIGGER_DATA_OUT_CLEAR_BIT, true},
+                 });
+
+  EXPECT_READ32(AES_STATUS_REG_OFFSET, 1);
+
+  EXPECT_WRITE32_SHADOWED(
+      AES_CTRL_SHADOWED_REG_OFFSET,
+      {{AES_CTRL_SHADOWED_OPERATION_OFFSET, AES_CTRL_SHADOWED_OPERATION_MASK},
+       {AES_CTRL_SHADOWED_MODE_OFFSET, AES_CTRL_SHADOWED_MODE_VALUE_AES_NONE},
+       {AES_CTRL_SHADOWED_KEY_LEN_OFFSET, AES_CTRL_SHADOWED_KEY_LEN_MASK}});
+
+  EXPECT_DIF_OK(dif_aes_reset(&aes_));
+}
+
+TEST_F(DifFunctionsTest, end) {
+  EXPECT_READ32(AES_STATUS_REG_OFFSET, 1);
+  EXPECT_READ32(AES_STATUS_REG_OFFSET, 1);
+  EXPECT_WRITE32_SHADOWED(AES_CTRL_SHADOWED_REG_OFFSET,
+                          {{AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, true}});
+
+  EXPECT_WRITE32(AES_TRIGGER_REG_OFFSET,
+                 {
+                     {AES_TRIGGER_KEY_IV_DATA_IN_CLEAR_BIT, true},
+                     {AES_TRIGGER_DATA_OUT_CLEAR_BIT, true},
+                 });
+
+  EXPECT_READ32(AES_STATUS_REG_OFFSET, 1);
+
+  EXPECT_DIF_OK(dif_aes_end(&aes_));
+}
+
+// Dif errors
+class DifBadArgError : public AesTestInitialized {
+ protected:
+  DifBadArgError() {
+    transaction.key_provider = kDifAesKeySoftwareProvided;
+    transaction.mode = kDifAesModeCbc;
+  }
+};
+
+TEST_F(DifBadArgError, start) {
+  EXPECT_DIF_BADARG(dif_aes_start(NULL, &transaction, &kKey_, &kIv_));
+  EXPECT_DIF_BADARG(dif_aes_start(&aes_, NULL, &kKey_, &kIv_));
+  EXPECT_DIF_BADARG(dif_aes_start(&aes_, &transaction, NULL, &kIv_));
+  EXPECT_DIF_BADARG(dif_aes_start(&aes_, &transaction, &kKey_, NULL));
+}
+
+TEST_F(DifBadArgError, reset) { EXPECT_DIF_BADARG(dif_aes_reset(NULL)); }
+
+TEST_F(DifBadArgError, end) { EXPECT_DIF_BADARG(dif_aes_end(NULL)); }
+
+TEST_F(DifBadArgError, LoadData) {
+  dif_aes_data_t data = {{0}};
```

Comment:
In C++, you can just write `{}` and it will initialize all members to zero (this unfortunately is not in C until C23).